code/modules/mob/living/simple_animal/bot/turtle/turtle_cartridge.dm

```diff
@@ -0,0 +1,280 @@
+			if("EQ")
+				return process_val(args[1]) == process_val(args[3])
+			if("NE")
+				return process_val(args[1]) != process_val(args[3])
+			if("OR")
+				return process_val(args[1]) || process_val(args[3])
+			if("AND")
+				return process_val(args[1]) && process_val(args[3])
+			else
+				throw_error("SYNTAX ERROR")
+
+/obj/item/disk/turtle_cartridge/proc/process_line()
+	if(!lines[line])
+		stop()
+		return
+
+	turtle?.speak("PROCESSING LINE [lines[line]]")
+
+	var/list/arguments = splittext(lines[line], " ")
+	if(!arguments.len) return
+
+	var/command = arguments[1]
+	arguments.Cut(1,2)
+	var/full_input = jointext(arguments, " ")
+
+	var/wait = 1
+
+	if (stack.len)
+		var/datum/stack_item/I = stack[stack.len]
+		if (command == "END")
+			if (istype(I, /datum/stack_item/cmd_while))
+				if(I.skipping)
+					stack.len --
+				else
+					if(process_expression(I.condition))
+						line = I.line
+					else
+						stack.len --
+			else if(istype(I, /datum/stack_item/cmd_if))
+				stack.len --
+			else
+				stack.len --
+
+		else if (I.skipping)
+			if(command == "IF" || command == "WHILE") //acount for if and while even if we're skipping so we match up ends
+				var/datum/stack_item/NI = new
+				NI.skipping = TRUE
+				stack += NI
+
+			line++
+			if (line > lines.len)
+				stop()
+			return
+
+	else if(command == "END")
+		throw_error("MISMATCHED END")
+
+	if(!running) return
+
+	switch(command)
+		if("MOVE", "GOTO", "WAIT")
+			if(arguments.len < 1) throw_error("INVALID ARG AMOUNT", arguments.len)
+			process_val(process_expression(arguments))
+		if("SET")
+			if(arguments.len < 2) throw_error("INVALID ARG AMOUNT", arguments.len)
+			if(!is_varname(arguments[1])) throw_error("INVALID VAR NAME", arguments[1])
+			if(!(arguments[1] in variables) && variables.len == memory_limit) throw_error("OUT OF MEMORY")
+			process_expression(arguments.Copy(2))
+		if("WHILE", "IF")
+			if(arguments.len < 1) throw_error("CONDITION NOT SUPPLIED")
+			process_expression(arguments)
+
+	if(!running) return //dont run any of this if we have stopped the program due to an error
+
+	switch(command)
+		if("SAY")
+			if(full_input) turtle?.speak(format_string(full_input))
+			wait = 10
+		if("BEEP")
+			turtle.beep("ping")
+			wait = 10
+		if("MOVE")
+			turtle?.Move(get_step(get_turf(turtle), process_expression(arguments)))
+			wait = 2
+		if("GOTO")
+			line = process_expression(arguments)
+		if("WAIT")
+			wait = process_expression(arguments)
+		if("SET")
+			variables[arguments[1]] = process_expression(arguments.Copy(2))
+		if("STOP")
+			stop()
+		if("IF")
+			var/datum/stack_item/cmd_if/I = new
+			I.condition = arguments
+			I.line = line
+			I.skipping = !process_expression(arguments)
+			stack += I
+		if("WHILE")
+			var/datum/stack_item/cmd_while/I = new
+			I.condition = arguments
+			I.line = line
+			I.skipping = !process_expression(arguments)
+			stack += I
+
+		if("#")
+			wait = 0 //basically ignore this line since it's a comment
+		if("END")
```

Comment:
🤔

Reply:
What

Reply:
Empty if block.

Reply:
So it doesn't throw command not found on end